src/machine.rs

```diff
@@ -420,6 +420,34 @@ impl<'tcx> PrimitiveLayouts<'tcx> {
     }
 }
 
+/// A thread's CPU affinity mask determines the set of CPUs on which it is eligible to run.
```

Comment:
What's the format for this? 1024 bits, each encoding one core? Why 1024?

Also, how does endianess factor in? From the comments below it seems like on little-endian, CPU N is encoded in element `N / 8`, bit `N % 8`, counting bits from least significant to most significant.

Reply:
Reading the man page, the internal kernel structure for this might be bigger. So an alternative would be to dynamically allocate something big enough for num_cpus.

I am fine with a static size as well, but then we should limit num_cpus so that, too. So there should be a `const MAX_CPUS = 1024` and  everything is computed from that.

Reply:
well I'm basing the implementation in this PR on the logic around `cpu_set_t` in the rust libc crate. Miri should at least be compatible with that I'd think.

I'll rework the logic with that constant.

Reply:
> well I'm basing the implementation in this PR on the logic around cpu_set_t in the rust libc crate. 

Then you need some FIXMEs in the implementation, since what this *should* implement is the logic of the glibc function with that name. It seems like the libc crate only exposes a part of that.

Reply:
well, the [man page](https://linux.die.net/man/3/cpu_alloc_size) does currently bound the cpu set at 1024 elements:

> The first available CPU on the system corresponds to a cpu value of 0, the next CPU corresponds to a cpu value of 1, and so on. The constant CPU_SETSIZE (currently 1024) specifies a value one greater than the maximum CPU number that can be stored in cpu_set_t.

is that not sufficient? I guess miri could assert on that constant being 1024 in practice?

Reply:
"the maximum CPU number that can be stored in cpu_set_t" is not the same as "the maximum number of CPUs supported by these APIs". The man page explicitly discusses this.

```
   Handling systems with large CPU affinity masks
       The underlying system calls (which represent CPU masks as bit masks of type unsigned long *) impose no restriction on the size of the CPU mask.  However, the cpu_set_t data type used by glibc has a fixed
       size of 128 bytes, meaning that the maximum CPU number that can be represented is 1023.  If the kernel CPU affinity mask is larger than 1024, then calls of the form:

           sched_getaffinity(pid, sizeof(cpu_set_t), &mask);

       fail with the error EINVAL, the error produced by the underlying system call for the case where the mask size specified in cpusetsize is smaller than the size of the affinity mask  used  by  the  kernel.
       (Depending on the system CPU topology, the kernel affinity mask can be substantially larger than the number of active CPUs in the system.)

       When working on systems with large kernel CPU affinity masks, one must dynamically allocate the mask argument (see CPU_ALLOC(3)).  Currently, the only way to do this is by probing for the size of the re‐
       quired mask using sched_getaffinity() calls with increasing mask sizes (until the call does not fail with the error EINVAL).

       Be aware that CPU_ALLOC(3) may allocate a slightly larger CPU set than requested (because CPU sets are implemented as bit masks allocated in units of sizeof(long)).  Consequently, sched_getaffinity() can
       set  bits  beyond  the requested allocation size, because the kernel sees a few additional bits.  Therefore, the caller should iterate over the bits in the returned set, counting those which are set, and
       stop upon reaching the value returned by CPU_COUNT(3) (rather than iterating over the number of bits requested to be allocated).
```